### PR TITLE
Remove circular reference of study.

### PR DESCRIPTION
**Pull request by keisuke-umezawa**
## Motivation

To solve the issue 2 written in https://github.com/optuna/optuna/issues/1347#issuecomment-650852619.

Actually, in the previous code, study is never garbage collected and `_finalize` is never called, because weakref is used wrongly c.f. [note at the bottom](https://docs.python.org/3/library/weakref.html#weakref.finalize). So, in my understanding, `_finalize` is not needed. Also, [SQL alchemy documentation](https://docs.sqlalchemy.org/en/13/orm/contextual.html#thread-local-scope) says that we do not need to call scoped_session.remove().

> The scoped_session.remove() method, as always, removes the current Session associated with the thread, if any. However, one advantage of the threading.local() object is that if the application thread itself ends, the “storage” for that thread is also garbage collected. So it is in fact “safe” to use thread local scope with an application that spawns and tears down threads, without the need to call scoped_session.remove().

## Description of the changes

- [x] Remove weakref which make circular reference. c.f. [note at the bottom](https://docs.python.org/3/library/weakref.html#weakref.finalize)

## checks
```
import optuna

conn_str= "postgresql://localhost:5432/foo"

for i in range(101):  # 100 is the maximum allowed number of connections for postgresql by default.
    study = optuna.create_study(conn_str, study_name=f"test_{i}")
    study._storage._backend.delete_study(study._study_id)
```

The above code run correctly.

### CODE REVIEW

File: optuna/storages/_rdb/storage.py [hunk 1]
```diff
@@ -1115,16 +1112,6 @@ def remove_session(self) -> None:
 
         self.scoped_session.remove()
 
-    def _finalize(self) -> None:
-
-        # This destructor calls remove_session to explicitly close the DB connection. We need this
-        # because DB connections created in SQLAlchemy are not automatically closed by reference
-        # counters, so it is not guaranteed that they are released by correct threads (for more
-        # information, please see the docstring of remove_session).
```
Comment (keisuke-umezawa) [inline optuna/storages/_rdb/storage.py:1123]: However, this comment says that `_finalize` is needed 🤔 Even though, it is never called.

**Comment by toshihikoyanase** [comment]
Thank you for your PR. As you and @hvy mentioned, `weakref.finalize`  should be fixed.

We have two issues about the finalization of `RDBStorage`, i.e., #938 and #1347. I roughly examined the reproducible code and this PR seems to resolve them. 

But, I'm not sure if we can just remove the finalization code. IMO, DB sessions should be closed somewhere before destroying the `RDBStorage` instances. Do you think the current implementation does not require it? 

Reproducible code for #938

```console
pip install jupyter
```

```python
import optuna

# Define a simple 2-dimensional objective function whose minimum value is -1 when (x, y) = (0, -1).
def objective(trial):
    x = trial.suggest_uniform('x', -100, 100)
    y = trial.suggest_categorical('y', [-1, 0, 1])
    return x**2 + y


if __name__ == '__main__':
    # Let us minimize the objective function above.
    print('Running 10 trials...')
    study = optuna.create_study(storage="sqlite:///db.sqlite3")
    study.optimize(objective, n_trials=10)
    print('Best value: {} (params: {})\n'.format(study.best_value, study.best_params))
```

Reproducible code for #1347

```console
docker run -it --rm --name postgres-test -e POSTGRES_PASSWORD=test -p 15432:5432 -d postgres
```

```python
from multiprocessing import Pool
import optuna
import random

conn_str = "postgres+psycopg2://postgres:test@localhost:15432/postgres"

def obj_func(trial:optuna.Trial,arg1:float):
    x=trial.suggest_float('x',0,100)
    return arg1*(x-3.3)**2

def task(arg1):
    study:optuna.Study=optuna.create_study(conn_str,study_name=f'test_{arg1}')
    study.optimize(lambda x:obj_func(x,arg1),n_trials=10)
    return study.best_trial

optuna.logging.disable_default_handler()
optuna.logging.set_verbosity(optuna.logging.ERROR)

# Create a dummy study to create DB tables.
optuna.create_study(conn_str)

task_ids = list(range(100))
random.shuffle(task_ids)

with Pool(maxtasksperchild=10) as pool:
    res=pool.map(task, task_ids)
```

**Comment by keisuke-umezawa** [comment]
@toshihikoyanase Thank you for the comment. I understand the context of introducing weakref. I will check #938, and reply your comment later.

**Comment by keisuke-umezawa** [comment]
I added the explanation for `remove` on the top comment.

[SQL alchemy documentation](https://docs.sqlalchemy.org/en/13/orm/contextual.html#thread-local-scope) says that we do not need to call scoped_session.remove().

> The scoped_session.remove() method, as always, removes the current Session associated with the thread, if any. However, one advantage of the threading.local() object is that if the application thread itself ends, the “storage” for that thread is also garbage collected. So it is in fact “safe” to use thread local scope with an application that spawns and tears down threads, without the need to call scoped_session.remove().

**Comment by keisuke-umezawa** [comment]
sqlalchemy's changes related to scoped session:

no change in documentation recently: 
https://github.com/sqlalchemy/sqlalchemy/blame/50417cf758508e970f3fccc99af6a501ceac8eb2/doc/build/orm/contextual.rst

no significant change in code recently: 
https://github.com/sqlalchemy/sqlalchemy/blame/master/lib/sqlalchemy/orm/scoping.py
https://github.com/sqlalchemy/sqlalchemy/blame/d55971119aa48590416193b8b0e0f54aa0e97c82/lib/sqlalchemy/util/_collections.py#L1017

**Review by HideakiImamura — APPROVED** [review]
Thanks for the PR and the detailed explanation. The removal of the `_finalize` method seems reasonable to me. LGTM!

**Comment by keisuke-umezawa** [comment]
@toshihikoyanase 
I surveyed materials related to this problem and summarized them on the above. So far, I can not reproduce the situation of b12b05d... Do you have any ideas to reproduce the situation?

**Comment by toshihikoyanase** [comment]
@keisuke-umezawa Thank you for your investigation. 

I guess #1628 may resolve the root cause of the issue mentioned in the `RDBStorage.remove_session` comment. https://github.com/optuna/optuna/blob/1acac077b3cf9ea581319cd8b95d0f83f089ce23/optuna/storages/_rdb/storage.py#L1084-L1097
It ensures that each session calls `commit` or `rollback` at the end of the storage methods and keeps the transaction clean. So, threads can close the sessions without calling `rollback`.

If we don't call the `session.commit()` or `session.rollback()` intentionally, we can see `sqlite3.ProgrammingError` at the end of program execution as follows:

```python
import optuna

def error_callback(study, trial):
    session = study._storage._backend.scoped_session()
    if trial.number > 5:
        try:
            optuna.storages._rdb.models.StudyModel.find_or_raise_by_name("ho", session)
        finally:
            # session.commit()
            pass

def objective(trial):
    return trial.suggest_float("x", 0, 1)

study = optuna.create_study(storage="sqlite:///bar.db")
try:
    study.optimize(objective, n_trials=10, callbacks=[error_callback], n_jobs=4)
except:
    pass
```

```console
sqlite3.ProgrammingError: SQLite objects created in a thread can only be used in that same thread. The object was created in thread id 140345088288512 and this is thread id 140345489463104.
Exception closing connection <sqlite3.Connection object at 0x7fa4a3228030>
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/pool/base.py", line 697, in _finalize_fairy
    fairy._reset(pool)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/pool/base.py", line 893, in _reset
    pool._dialect.do_rollback(self)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 543, in do_rollback
    dbapi_connection.rollback()
sqlite3.ProgrammingError: SQLite objects created in a thread can only be used in that same thread. The object was created in thread id 140345088288512 and this is thread id 140345489463104.
```

In summary, I think we can accept this change.

**Review by toshihikoyanase — APPROVED** [review]
LGTM. Thank you for resolving the long-standing issue.
